### DISCUSSION TITLE
CronTrigger with configurable "random backoff time"

### DISCUSSION THREAD

**Discussion by andyolivares** [quartznet/quartznet#1482] [Q&A]
Hi:

I'm wondering if there is a way to set up a CronTrigger with a configurable "random backoff time". I really don't know what's the correct term here but here's an example:

Let's say I have a CronTrigger to fire every hour, so for example, it should trigger at:

- 0:00 AM
- 1:00 AM
- 2:00 AM
- ...

But what I need is it to actually fire with a random +- 5 minutes (configurable):

- 0:03 AM
- 0:57 AM
- 2:01 AM
- ...

Is that possible? My use case is that I have many jobs with the same configured CronTrigger (across many computers) but I don't want to have all of them calling my backend API at the same time.

Thank you for your help :)

**Comment by lahma** [comment]
Can't say how to do it out-of-the-box, but you could for example create random sleep in your job (would only delay so no before-fire behavior)  or you could use this cron-job to create randomly scheduled one-time triggers to run the actual job.

Of course if you have a cluster and want only one instance of job (same job key) to be run at the same time you could just mark the job with `[DisallowConcurrentExecution]` attribute.

**Reply by andyolivares** [reply]
> you could use this cron-job to create randomly scheduled one-time triggers to run the actual job

That's what I thought at the beginning but wanted to know if there is any other "more elegant" way :)

What about inheriting CronTriggerImpl with that functionalify? Will overriding GetNextFireTimeUtc() with a random value added will do the trick?

**Reply by lahma** [reply]
Overriding has the problem that it should return the same value even if called multiple times during Quartz processing. Might be hard to memoize that.

**Reply by andyolivares** [reply]
What if random value is calculated on SetNextFireTimeUtc() and then use that value onwards (at least while SetNextFireTimeUtc is not called again)?

**Reply by lahma** [reply]
Well please try out how it works, Quartz wasn't designed to do random scheduling, more deterministic. If it's all about not hammering an API also consider a design where job is configured to disallow concurrent execution and other nodes wait for their time to trigger.

**Reply by andyolivares** [reply]
> also consider a design where job is configured to disallow concurrent execution and other nodes wait for their time to trigger

It is my understading that Jobs/Triggers are managed and shared between all nodes, right? So if a trigger returns a particular value when calling GetNextFireTimeUtc() should be the same for all nodes?

In other words, who calls GetNextFireTimeUtc() and makes sure the trigger data will be consistent between nodes?

I can see a NEXT_FIRE_TIME field in the DB, so I assume that the value of this field has something to do with the value returned by GetNextFireTimeUtc() being shared between nodes.

Sorry for all the questions, just want to make sure I understand things correctly before attempting any modifications :)

Thank you so much!

**Reply by lahma** [reply]
Yes, in a cluster, configuration is shared (types are shared as long as they work between nodes, DLL needs to be present, otherwise exceptions). The  `GetNextFireTimeUtc` is being called to determine next fire and it's stored in DB for polling - for selecting next trigger to process. This however, does not guarantee that `GetNextFireTimeUtc` is called only once in the logic, so it should return stable values.

What I was trying to say that if I have C# code that calls the method, it should return same value when called consecutively, like:

```csharp
Assert.Equals(trigger.GetNextFireTimeUtc(), trigger.GetNextFireTimeUtc());
```

In other words, `GetNextFireTimeUtc` is not the best place to have randomness. But you can try what tweaks you can make, I cannot guarantee it will work from version to version if there's randomness in play.

**Reply by andyolivares** [reply]
Ok. I understand now. Will try something and let you know.
